cratedb_toolkit/wtf/library.py

```diff
@@ -0,0 +1,587 @@
+                    DATE_FORMAT('%Y-%m-%dT%H:%i:%s.%f', ended) AS ended,
+                    classification, stmt, username, node
+                FROM
+                    sys.jobs_log
+                WHERE
+                    stmt NOT LIKE '%sys.%' AND
+                    stmt NOT LIKE '%information_schema.%'
+                ORDER BY ended DESC
+                LIMIT {limit};
+            """,
+        )
+
+    class Replication:
+        """
+        Information about logical replication.
+        """
+
+        # https://github.com/crate/crate/blob/master/docs/admin/logical-replication.rst#monitoring
+        subscriptions = """
+        SELECT s.subname, s.subpublications, sr.srrelid::text, sr.srsubstate, sr.srsubstate_reason
+        FROM pg_subscription s
+        JOIN pg_subscription_rel sr ON s.oid = sr.srsubid
+        ORDER BY s.subname;
+        """
+
+    class Resources:
+        """
+        About system resources.
+        """
+
+        # TODO: Needs templating.
+        column_cardinality = """
+        SELECT tablename, attname, n_distinct
+        FROM pg_stats
+        WHERE schemaname = '...'
+        AND tablename IN (...)
+        AND attname IN (...);
+        """
+
+        file_descriptors = """
+        SELECT
+            name AS node_name,
+            process['open_file_descriptors'] AS "open_file_descriptors",
+            process['max_open_file_descriptors'] AS max_open_file_descriptors
+        FROM sys.nodes
+        ORDER BY node_name;
+        """
+
+        heap_usage = """
+        SELECT
+            name AS node_name,
+            heap['used'] / heap['max']::DOUBLE AS heap_used
+        FROM sys.nodes
+        ORDER BY node_name;
+        """
+
+        tcp_connections = """
+        SELECT
+            name AS node_name,
+            connections
+        FROM sys.nodes
+        ORDER BY node_name;
+        """
+
+        # TODO: Why "14"? Is it about only getting information about the `write` thread pool?
+        thread_pools = """
+        SELECT
+            name AS node_name,
+            thread_pools[14]['queue'],
+            thread_pools[14]['active'],
+            thread_pools[14]['threads']
+        FROM sys.nodes
+        ORDER BY node_name;
+        """
+
+    class Settings:
+        """
+        Reflect cluster settings.
+        """
+
+        info = """
+        SELECT
+            name,
+            master_node,
+            settings['cluster']['routing']['allocation']['cluster_concurrent_rebalance']
+                AS cluster_concurrent_rebalance,
+            settings['indices']['recovery']['max_bytes_per_sec'] AS max_bytes_per_sec
+        FROM sys.cluster
+        LIMIT 1;
+        """
+
+    class Shards:
+        """
+        Information about shard / node / table / partition allocation and rebalancing.
+        """
+
+        # https://cratedb.com/docs/crate/reference/en/latest/admin/system-information.html#example
+        # TODO: Needs templating.
```

Comment:
I agree. On the other hand, this is just an SQL statement slotted into the library, but it will not get used on behalf of any implementation yet, so it's not directly possible to slap a `raise NotImplementedError()` onto it.